contracts/chain-adapters/ForwarderBase.sol

```diff
@@ -0,0 +1,119 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity ^0.8.0;
+
+import { UUPSUpgradeable } from "@openzeppelin/contracts-upgradeable/proxy/utils/UUPSUpgradeable.sol";
+import { EIP712CrossChainUpgradeable } from "../upgradeable/EIP712CrossChainUpgradeable.sol";
+
+abstract contract ForwarderBase is UUPSUpgradeable, EIP712CrossChainUpgradeable {
+    // L3 address of the recipient of fallback messages and tokens.
+    address public l3SpokePool;
+
+    // L1 address of the contract which can relay messages to the l3SpokePool contract and update this proxy contract.
+    address public crossDomainAdmin;
+
+    event TokensForwarded(address indexed l2Token, uint256 amount);
+    event MessageForwarded(address indexed target, bytes message);
+    event SetXDomainAdmin(address indexed crossDomainAdmin);
+    event SetL3SpokePool(address indexed l3SpokePool);
+
+    error InvalidCrossDomainAdmin();
+    error InvalidL3SpokePool();
+
+    /*
+     * @dev All functions with this modifier must revert if msg.sender != CROSS_DOMAIN_ADMIN, but each L2 may have
+     * unique aliasing logic, so it is up to the forwarder contract to verify that the sender is valid.
+     */
+    modifier onlyAdmin() {
+        _requireAdminSender();
+        _;
+    }
+
+    /**
+     @notice Constructs the Forwarder contract.
+     @dev _disableInitializers() restricts anybody from initializing the implementation contract, which if not done, 
+     * may disrupt the proxy if another EOA were to initialize it.
+     */
+    constructor() {
+        _disableInitializers();
+    }
+
+    /**
+     * @notice Initializes the forwarder contract.
+     * @param _l3SpokePool L3 address of the contract which will receive messages and tokens which are temporarily
+     * stored in this contract on L2.
+     * @param _crossDomainAdmin L1 address of the contract which can send root bundles/messages to this forwarder contract.
+     * In practice, this is the hub pool.
+     */
+    function __Forwarder_init(address _l3SpokePool, address _crossDomainAdmin) public onlyInitializing {
+        __UUPSUpgradeable_init();
+        __EIP712_init("ACROSS-V2", "1.0.0");
```

Comment:
What's the use of EIP-712 here? We have it in the SpokePool to permit depositors to update their deposit details but I don't see how it's related to transport of tokens and messages to an L3.

Reply:
Ah I didn't realize that is all it was used for: https://github.com/across-protocol/contracts/pull/609/commits/fe66b780fa9c9e63e79f15fe7827bd2c6b86e3bb